data-generator/src/main/java/com/github/streamshub/kafka/data/generator/DataGenerator.java

```diff
@@ -0,0 +1,64 @@
+package com.github.streamshub.kafka.data.generator;
+
+import org.apache.avro.generic.GenericData;
+import org.apache.avro.generic.GenericRecord;
+import org.apache.kafka.clients.producer.KafkaProducer;
+import org.apache.kafka.clients.producer.Producer;
+import org.apache.kafka.clients.producer.ProducerRecord;
+
+import java.util.Map;
+import java.util.concurrent.ExecutionException;
+import java.util.function.Supplier;
+
+public class DataGenerator implements Runnable {
+    final String bootstrapServers;
+    final Data data;
+
+    public DataGenerator(String bootstrapServers, Data data) {
+        this.bootstrapServers = bootstrapServers;
+        this.data = data;
+    }
+
+    @Override
+    public void run() {
+        if (Boolean.parseBoolean(System.getenv("USE_APICURIO_REGISTRY"))) {
+            String registryUrl = System.getenv("REGISTRY_URL");
+            Producer<String, Object> producer = new KafkaProducer<>(KafkaClientProps.avro(bootstrapServers, registryUrl));
+            send(producer, data.topic(), () -> {
+                Map<String, String> generatedData = data.generate();
+                GenericRecord avroRecord = new GenericData.Record(data.schema());
+                generatedData.forEach(avroRecord::put);
+                return avroRecord;
+            });
+        } else {
+            Producer<String, String> producer = new KafkaProducer<>(KafkaClientProps.csv(bootstrapServers));
+            send(producer, data.topic(), () -> String.join(",", data.generate().values()));
+        }
+    }
+
+    private <V> void send(Producer<String, V> producer, String topic, Supplier<V> valueSupplier) {
+        try (producer) {
+
+            while (true) {
+                String key = String.valueOf(System.currentTimeMillis());
+                V value = valueSupplier.get();
+                ProducerRecord<String, V> record =
+                        new ProducerRecord<>(
+                                topic,
+                                key,
+                                valueSupplier.get());
+
+                producer.send(record).get();
```

Comment:
On my machine one of the sending threads crashed with an apicurio exception when it tried to send the first message. This lead to the job as a whole producing no output because each source needs events flowing through it to drive watermarking.

Unfortunately I lost the logs when I deleted the pod, but it looked like an async error due to two threads trying to register schema's simultaneously. 

Perhaps the while loop could be more lenient, running a try-catch-finally where we do the 3 second sleep in the finally. The catch would log the error but let the thread try again indefinitely. WDYT?